pkg/storage/object.go

```diff
@@ -0,0 +1,314 @@
+	if !s.BucketExists(bucket) {
+		return "", ErrBucketNotFound
+	}
+
+	objectDir, err := s.safePath(bucket, key)
+	if err != nil {
+		return "", err
+	}
+
+	// Create object directory
+	if err := os.MkdirAll(objectDir, 0755); err != nil {
+		return "", err
+	}
+
+	dataPath := filepath.Join(objectDir, dataFile)
+	metaPath := filepath.Join(objectDir, metaFile)
+
+	// Create temp file in the object directory
+	tmpFile, err := os.CreateTemp(objectDir, ".tmp-*")
+	if err != nil {
+		return "", err
+	}
+	defer os.Remove(tmpFile.Name())
+
+	// Calculate MD5 while writing
+	hash := md5.New()
+	writer := io.MultiWriter(tmpFile, hash)
+
+	if _, err := io.Copy(writer, data); err != nil {
+		tmpFile.Close()
+		return "", err
+	}
+	tmpFile.Close()
+
+	// Move temp file to data file
+	if err := os.Rename(tmpFile.Name(), dataPath); err != nil {
+		return "", err
+	}
+
+	// Store metadata
+	etag := hex.EncodeToString(hash.Sum(nil))
+	metadata := map[string]string{
+		"Content-Type": contentType,
+		"ETag":         etag,
+	}
+	if err := s.saveMetadata(metaPath, metadata); err != nil {
+		return "", err
+	}
+
+	return etag, nil
+}
+
+// GetObject retrieves an object
+func (s *Storage) GetObject(bucket, key string) (io.ReadCloser, *ObjectInfo, error) {
+	if !s.BucketExists(bucket) {
+		return nil, nil, ErrBucketNotFound
+	}
+
+	objectDir, err := s.safePath(bucket, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dataPath := filepath.Join(objectDir, dataFile)
+	metaPath := filepath.Join(objectDir, metaFile)
+
+	// Check if object exists
+	file, err := os.Open(dataPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil, ErrObjectNotFound
+		}
+		return nil, nil, err
+	}
+
+	// Get file info
+	fileInfo, err := file.Stat()
+	if err != nil {
+		file.Close()
+		return nil, nil, err
+	}
+
+	// Load metadata
+	metadata, err := s.loadMetadata(metaPath)
+	if err != nil {
+		file.Close()
+		return nil, nil, err
+	}
+
+	info := &ObjectInfo{
+		Key:          key,
+		Size:         fileInfo.Size(),
+		ETag:         metadata["ETag"],
+		LastModified: fileInfo.ModTime(),
+		ContentType:  metadata["Content-Type"],
+	}
+
+	if info.ContentType == "" {
+		info.ContentType = "application/octet-stream"
+	}
+
+	return file, info, nil
+}
+
+// DeleteObject deletes an object
+func (s *Storage) DeleteObject(bucket, key string) error {
+	if !s.BucketExists(bucket) {
+		return ErrBucketNotFound
+	}
+
+	objectDir, err := s.safePath(bucket, key)
+	if err != nil {
+		return err
+	}
+
+	// Check if object exists
+	if _, err := os.Stat(filepath.Join(objectDir, dataFile)); os.IsNotExist(err) {
+		return ErrObjectNotFound
+	}
+
+	// Remove entire object directory
+	return os.RemoveAll(objectDir)
+}
+
+// ListObjects lists objects in a bucket with optional prefix and delimiter
+func (s *Storage) ListObjects(bucket, prefix, delimiter string, maxKeys int) ([]ObjectInfo, []string, error) {
```

Comment:
cyclomatic complexity 16 of func `(*Storage).ListObjects` is high (> 15) (gocyclo)